### PR TITLE
fix(conversation): use functionMap for custom handler IFunction reference

### PR DESCRIPTION
**Pull request by atierian**
> Related data-schema PR https://github.com/aws-amplify/amplify-api-next/pull/361

## Description of changes

Updates ConversationTransformer to allow `functionMap?: record<string, lambda.IFunction>` to be passed. This is done to properly reference a custom conversation handler defined via `defineConversationHandlerFunction` in a Gen2 backend.

## Description of how you validated changes
- New test case asserting `functionMap` value is used for Lambda data source.
## Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] ~E2E test run linked~
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] ~Relevant documentation is changed or added (and PR referenced)~
- [ ] ~New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies~
- [ ] ~Any CDK or CloudFormation parameter changes are called out explicitly~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-conversation-transformer/package.json [hunk 1]
```diff
@@ -37,7 +37,8 @@
     "immer": "^9.0.12"
   },
   "devDependencies": {
-    "@aws-amplify/graphql-transformer-test-utils": "1.0.3"
+    "@aws-amplify/graphql-transformer-test-utils": "1.0.3",
+    "esbuild": "^0.24.0"
```
Comment (atierian) [inline packages/amplify-graphql-conversation-transformer/package.json:41]: required for test cases

File: dependency_licenses.txt [hunk 2]
```diff
@@ -14486,6 +14486,32 @@ SOFTWARE.
 
 -----
 
+The following software may be included in this product: esbuild. A copy of the source code may be downloaded from git+https://github.com/evanw/esbuild.git. This software contains the following license and notice below:
```
Comment (atierian) [inline dependency_licenses.txt:14489]: This is from adding `esbuild` as a dev dependency in the conversation transformer.

File: packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts [hunk 3]
```diff
@@ -44,6 +45,35 @@ describe('ConversationTransformer', () => {
...
+      expect(out).toBeDefined();
+
+      const expectedCustomHandlerArn = out.rawRootStack.resolve(customHandler.functionArn);
+      const conversationLambdaStackName = `${toUpper(routeName)}ConversationDirectiveLambdaStack`;
+      const conversationLambdaDataSourceName = `Fn${routeName}LambdaDataSource`;
+      const conversationLambdaDataSourceFunctionArnRef =
+        out.stacks[conversationLambdaStackName].Resources?.[conversationLambdaDataSourceName].Properties.LambdaConfig.LambdaFunctionArn.Ref;
+      const lambdaDataSourceFunctionArn =
+        out.rootStack.Resources?.[conversationLambdaStackName].Properties?.Parameters?.[conversationLambdaDataSourceFunctionArnRef];
+      expect(lambdaDataSourceFunctionArn).toEqual(expectedCustomHandlerArn);
```
Comment (atierian) [inline packages/amplify-graphql-conversation-transformer/src/__tests__/amplify-graphql-conversation-transformer.test.ts:75]: I welcome suggestions for a more succinct way to do this, like using CDK test helpers -- I was unable to find a better way.

**Review by sundersc — APPROVED** [review]

File: packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts [hunk 4]
```diff
@@ -101,28 +103,18 @@ export class ConversationResolverGenerator {
    * @param functionStack - The CDK stack to add the function to
```
Comment (palpatim) [inline packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts:103]: nit: delete
```suggestion
```

**Review by palpatim — APPROVED** [review]

File: packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts (hunk 4 above)
Comment (atierian) [reply packages/amplify-graphql-conversation-transformer/src/transformer-steps/conversation-resolver-generator.ts:103]: thanks. will remove in follow up.

**Pull request merged**